File: 골드 4 (3문제)/1043_거짓말.py
# 설명 링크 : https://moon-gd.tistory.com/133
import sys
q_input = sys.stdin.readline  # 빠른 입력

N, M = map(int, q_input().split())  # 입력 : 사람, 파티 수

# 입력 : 진실을 아는 사람
know_true = [False for _ in range(N + 1)]
for index in list(map(int, q_input().split()))[1:]:
    know_true[index] = True

# 입력 : 파티 참석 인원 정보
parties = []
for _ in range(M):
    parties.append(list(map(int, q_input().split()))[1:])


# 진실을 새롭게 알게 된 사람이 있는지를 알려주는 함수
# 즉, return True : 진실을 알게 된 사람이 있다
def update():
    changed = False

    for party in parties:
        flag = False
        for person in party:
            # 특정 파티에서 진실을 아는 사람이 있는 경우
            if know_true[person]:
                # flag, changed 갱신 후 break
                flag = True
                changed = True
                break

        # flag 변수가 활성화 된 경우
        if flag:
            # 해당 파티 인원 모두 진실을 알도록 갱신
            for person in party:
                know_true[person] = True

            # 중복 탐색을 방지하기 위해
            # 갱신이 끝나면 해당 파티 정보 삭제
            parties.remove(party)

    return changed


while True:
    # update()를 더 이상 수행하지 않아도 되는 경우 while 탈출
    if update() is not True:
        break

# 정답 출력
count = len(parties)
print(count)

File: 골드 4 (3문제)/1043_거짓말.c
// 설명 링크 : https://moon-gd.tistory.com/133
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
#define MAX 51

// 사람 수, 파티 수, 진실을 아는 사람 수
int person_number, party_number, know_true_number;

int party[MAX][MAX];

// 진실을 아는 사람 정보
// 값이 1인 경우 : 진실을 아는 사람
// 값이 0인 경우 : 진실을 모르는 사람
int know_true[MAX];

// 진실을 알게 된 사람 정보를 갱신해주는 함수
// 새롭게 진실을 알게 된 사람이 생긴 경우 : return 1
// 이전과 같은 경우 : return 0
int update();

int main() {
	// 입력 : 사람 수, 파티 수, 진실을 아는 사람 수
	scanf("%d %d %d", &person_number, &party_number, &know_true_number);

	// 입력 : 진실을 아는 사람 정보
	int temp_num;
	for (int i = 0; i < know_true_number; i++) {
		scanf("%d", &temp_num);
		know_true[temp_num] = 1;
	}

	// 입력 : 파티 인원 수, 파티 참여 인원
	for (int i = 0; i < party_number; i++) {
		scanf("%d", &party[i][0]);

		for (int j = 1; j <= party[i][0]; j++) {
			scanf("%d", &party[i][j]);
		}
	}

	// 변경이 발생하지 않을 경우 while문 탈출
	do {
	} while (update());

	// 결과 계산
	int count = 0;
	for (int i = 0; i < party_number; i++) {
		// 파티 인원이 0이 아니라면? update() 함수의 영향을 받지 않았음을 의미
		// 이는 파티 인원이 모두 진실을 모른다는 것을 의미!
		// 따라서 count에 1을 더해준다
		if (party[i][0]) {
			count++;
		}
	}

	// 결과 출력
	printf("%d", count);

	return 0;
}

// 진실을 알게 된 사람 정보를 갱신해주는 함수
// 새롭게 진실을 알게 된 사람이 생긴 경우 : return 1
// 이전과 같은 경우 : return 0
int update() {
	int changed = 0;
	for (int i = 0; i < party_number; i++) {
		int flag = 0;
		for (int j = 1; j <= party[i][0]; j++) {
			if (know_true[party[i][j]]) {
				flag = 1;
				break;
			}
		}

		if (flag) {
			for (int j = 1; j <= party[i][0]; j++) {
				if (know_true[party[i][j]] == 0) {
					know_true[party[i][j]] = 1;
					changed = 1;
				}
			}
			party[i][0] = 0;
		}
	}

	return changed;
}